File: ts-src/src/local-tool-shortcuts.ts
export type LocalToolShortcut =
  | { toolName: 'list_files'; input: { path?: string } }
  | { toolName: 'grep_files'; input: { pattern: string; path?: string } }
  | { toolName: 'read_file'; input: { path: string } }
  | { toolName: 'write_file'; input: { path: string; content: string } }
  | { toolName: 'modify_file'; input: { path: string; content: string } }
  | { toolName: 'edit_file'; input: { path: string; search: string; replace: string } }
  | {
      toolName: 'patch_file'
      input: {
        path: string
        replacements: Array<{ search: string; replace: string; replaceAll?: boolean }>
      }
    }
  | { toolName: 'run_command'; input: { command: string; args?: string[]; cwd?: string } }

export function parseLocalToolShortcut(input: string): LocalToolShortcut | null {
  if (input.startsWith('/ls')) {
    const dir = input.replace('/ls', '').trim()
    return {
      toolName: 'list_files',
      input: dir ? { path: dir } : {},
    }
  }

  if (input.startsWith('/grep ')) {
    const payload = input.slice('/grep '.length).trim()
    const [pattern, searchPath] = payload.split('::')
    if (!pattern?.trim()) return null
    return {
      toolName: 'grep_files',
      input: {
        pattern: pattern.trim(),
        path: searchPath?.trim() || undefined,
      },
    }
  }

  if (input.startsWith('/read ')) {
    const filePath = input.slice('/read '.length).trim()
    if (!filePath) return null
    return {
      toolName: 'read_file',
      input: { path: filePath },
    }
  }

  if (input.startsWith('/write ')) {
    const payload = input.slice('/write '.length)
    const splitAt = payload.indexOf('::')
    if (splitAt === -1) return null
    return {
      toolName: 'write_file',
      input: {
        path: payload.slice(0, splitAt).trim(),
        content: payload.slice(splitAt + 2),
      },
    }
  }

  if (input.startsWith('/modify ')) {
    const payload = input.slice('/modify '.length)
    const splitAt = payload.indexOf('::')
    if (splitAt === -1) return null
    return {
      toolName: 'modify_file',
      input: {
        path: payload.slice(0, splitAt).trim(),
        content: payload.slice(splitAt + 2),
      },
    }
  }

  if (input.startsWith('/edit ')) {
    const payload = input.slice('/edit '.length)
    const [targetPath, search, replace] = payload.split('::')
    if (!targetPath || search === undefined || replace === undefined) {
      return null
    }
    return {
      toolName: 'edit_file',
      input: {
        path: targetPath.trim(),
        search,
        replace,
      },
    }
  }

  if (input.startsWith('/cmd ')) {
    const payload = input.slice('/cmd '.length).trim()
    const splitAt = payload.indexOf('::')
    const commandText = splitAt === -1 ? payload : payload.slice(splitAt + 2).trim()
    const commandCwd = splitAt === -1 ? undefined : payload.slice(0, splitAt).trim()
    const parts = commandText.split(/\s+/)
    const [command, ...args] = parts
    if (!command) return null
    return {
      toolName: 'run_command',
      input: { command, args, cwd: commandCwd || undefined },
    }
  }

  if (input.startsWith('/patch ')) {
    const payload = input.slice('/patch '.length)
    const [targetPath, ...ops] = payload.split('::')
    if (!targetPath?.trim() || ops.length < 2 || ops.length % 2 !== 0) {
      return null
    }

    const replacements = []
    for (let i = 0; i < ops.length; i += 2) {
      replacements.push({
        search: ops[i] ?? '',
        replace: ops[i + 1] ?? '',
      })
    }

    return {
      toolName: 'patch_file',
      input: {
        path: targetPath.trim(),
        replacements,
      },
    }
  }

  return null
}

File: ts-src/py-src/minicode/local_tool_shortcuts.py
from __future__ import annotations


def parse_local_tool_shortcut(user_input: str) -> dict | None:
    if user_input.startswith("/ls"):
        directory = user_input.replace("/ls", "", 1).strip()
        return {"toolName": "list_files", "input": {"path": directory} if directory else {}}

    if user_input.startswith("/grep "):
        payload = user_input[len("/grep ") :].strip()
        pattern, _, search_path = payload.partition("::")
        if not pattern.strip():
            return None
        input_data = {"pattern": pattern.strip()}
        if search_path.strip():
            input_data["path"] = search_path.strip()
        return {"toolName": "grep_files", "input": input_data}

    if user_input.startswith("/read "):
        file_path = user_input[len("/read ") :].strip()
        return {"toolName": "read_file", "input": {"path": file_path}} if file_path else None

    if user_input.startswith("/write "):
        payload = user_input[len("/write ") :]
        target_path, separator, content = payload.partition("::")
        if not separator:
            return None
        return {
            "toolName": "write_file",
            "input": {"path": target_path.strip(), "content": content},
        }

    if user_input.startswith("/modify "):
        payload = user_input[len("/modify ") :]
        target_path, separator, content = payload.partition("::")
        if not separator:
            return None
        return {
            "toolName": "modify_file",
            "input": {"path": target_path.strip(), "content": content},
        }

    if user_input.startswith("/edit "):
        payload = user_input[len("/edit ") :]
        parts = payload.split("::")
        if len(parts) != 3:
            return None
        target_path, search, replace = parts
        return {
            "toolName": "edit_file",
            "input": {"path": target_path.strip(), "search": search, "replace": replace},
        }

    if user_input.startswith("/patch "):
        payload = user_input[len("/patch ") :]
        parts = payload.split("::")
        if len(parts) < 3 or len(parts) % 2 == 0:
            return None
        target_path, *ops = parts
        replacements = []
        for index in range(0, len(ops), 2):
            replacements.append({"search": ops[index], "replace": ops[index + 1]})
        return {
            "toolName": "patch_file",
            "input": {"path": target_path.strip(), "replacements": replacements},
        }

    if user_input.startswith("/cmd "):
        payload = user_input[len("/cmd ") :].strip()
        cwd, separator, command_text = payload.partition("::")
        text = command_text.strip() if separator else payload
        command_cwd = cwd.strip() if separator else None
        if not text:
            return None
        return {
            "toolName": "run_command",
            "input": {"command": text, "cwd": command_cwd or None},
        }

    return None
